asserts/asserts.go

```diff
@@ -218,11 +220,173 @@ func Decode(serializedAssertion []byte) (Assertion, error) {
 	}
 
-	if len(signature) == 0 {
-		return nil, fmt.Errorf("empty assertion signature")
+	return Assemble(headers, body, content, signature)
+}
+
+// Maximum assertion component sizes.
+const (
+	MaxBodySize      = 2 * 1024 * 1024
+	MaxHeadSize      = 128 * 1024
+	MaxSignatureSize = 128 * 1024
+)
+
+// Decoder parses a stream of assertions bundled by ensuring double newlines at the end of each assertion.
+type Decoder struct {
+	rd             io.Reader
+	initialBufSize int
+	b              *bufio.Reader
+	err            error
+	maxHeadSize    int
+	maxBodySize    int
+	maxSigSize     int
+}
+
+func newDecoder(r io.Reader, bufSize, maxHeadSize, maxBodySize, maxSigSize int) *Decoder {
+	return &Decoder{
+		rd:             r,
+		initialBufSize: bufSize,
+		b:              bufio.NewReaderSize(r, bufSize),
+		maxHeadSize:    maxHeadSize,
+		maxBodySize:    maxBodySize,
+		maxSigSize:     maxSigSize,
 	}
+}
 
-	return Assemble(headers, body, content, signature)
+const defaultDecoderButSize = 4096
+
+// NewDecoder returns a Decoder to parse the stream of assertions from the reader.
+func NewDecoder(r io.Reader) *Decoder {
+	return newDecoder(r, defaultDecoderButSize, MaxHeadSize, MaxBodySize, MaxSignatureSize)
+}
+
+func (d *Decoder) peek(size int) ([]byte, error) {
+	buf, err := d.b.Peek(size)
+	if err == bufio.ErrBufferFull {
+		rebuf, reerr := d.b.Peek(d.b.Buffered())
+		if reerr != nil {
+			panic(reerr)
+		}
+		mr := io.MultiReader(bytes.NewBuffer(rebuf), d.rd)
+		d.b = bufio.NewReaderSize(mr, (size/d.initialBufSize+1)*d.initialBufSize)
+		buf, err = d.b.Peek(size)
+	}
+	if err != nil && d.err == nil {
+		d.err = err
+	}
+	return buf, d.err
+}
+
+func (d *Decoder) readExact(size int) ([]byte, error) {
+	buf, err := d.peek(size)
+	d.b.Discard(len(buf))
+	if len(buf) == size {
+		return buf, nil
+	}
+	if err == io.EOF {
+		return buf, io.ErrUnexpectedEOF
+	}
+	return buf, err
+}
+
+func (d *Decoder) readUntil(delim []byte, maxSize int) ([]byte, error) {
+	last := 0
+	size := d.initialBufSize
+	for {
+		buf, err := d.peek(size)
+		if i := bytes.Index(buf[last:], delim); i >= 0 {
+			d.b.Discard(last + i + len(delim))
+			return buf[:last+i+len(delim)], nil
+		}
+		// report errors only once we have consumed what is buffered
+		if err != nil && len(buf) == d.b.Buffered() {
+			d.b.Discard(len(buf))
+			return buf, err
+		}
+		last = size - len(delim) + 1
+		size *= 2
+		if size > maxSize {
+			return nil, fmt.Errorf("max size exceeded")
+		}
+	}
+}
+
+// Decode parses the next assertion from the stream.
+// It returns the error io.EOF at the end of a well-formed stream.
+func (d *Decoder) Decode() (Assertion, error) {
+	// read the headers and the nlnl separator after them
+	headAndSep, err := d.readUntil(nlnl, d.maxHeadSize)
+	if err != nil {
+		if err == io.EOF {
+			if len(headAndSep) != 0 {
+				return nil, io.ErrUnexpectedEOF
+			}
+			return nil, io.EOF
+		}
+		return nil, fmt.Errorf("error reading assertion head: %v", err)
```

Comment:
As a non-enforced but suggested convention, the code generally reads more easily as
```
if err == io.EOF {
    // Do something and return.
}
if err != nil {
    // Do something else and return.
}
```

Besides the better alignment making clear what the possibilities are, the first branch condition stays closer to its logic.